README
			PSUEDO TTY LAYERS PACKAGE

INTRODUCTION
	This is the pseudo-tty version of AT&T's layers software package.
It was originally written for 4.3BSD, especially Suns.

	This package was originally written by Keith Muller at the
University Of California San Diego and has had numerous enhancements by
Dave Dykstra at AT&T.  The package is public domain.

	The Keith Muller package had been tested on numerous versions of
BSD unix and was known to function properly on the following: Vax, Sun, CCI,
Unisys, ICL, Harris, Alliant, FPS and Gould.  However, many of them have
not been tested in a long time.

	The Dave Dykstra version has been changed to require even fewer
BSDisms.  Theoretically the only significant features a Unix should really
need is pseudo ttys and the winsize/TIOCSWINSZ/TIOCGWINSZ feature for
keeping track of window sizes.  Systems that have now been tried are
AT&T/NCR '386 SVR4.0, HP-PA 9.0, and AT&T/NCR Starserver FT (Mips-processor
based with a variant of RISC/os by Tandem).  See PORTING below for more
information if your Unix is not one of the supported types.


DIFFERENCES BETWEEN THE PSEUDO-TTY AND XT-DRIVER VERSIONS OF LAYERS
	The environment variable DMD does not have to be set in the
pseudo-tty layers.  If the environment variable DMD is not set, then the
compiled-in default path (typically /usr/local/dmdlayers, /usr/local/630,
or /usr/add-on/630) will be used.  You can use the DMD environment feature
as described in the ATT terminal manuals to place the software in a
different location.  Alternatively you can use a symbolic link at the
compiled-in location pointing to the location where you have installed it.
If you have source, a single change to the DMD macro in the top level
makefile, followed by a complete recompile will change the default path
used by all the installed binaries.

	Pseudo-tty layers no longer needs to be setuid-root.  Instead, one
or two small support programs are used.  One program handles moving entries
in the 'who' file (/etc/utmp) and the second handles granting ownership of
pseudo-ttys.  The first program may not need to be setuid-root if either
/etc/utmp is world-writable or you're on system that allows non-root
programs to call pututxline().  The second program may not be necessary at
all if your system has a grantpt() system call, but if it is necessary it
needs to be installed setuid-root to work.  Layers looks for these support
programs in two places: $DMD/lib/layersys and $DMDSYS/lib/layersys.  Both
$DMD and $DMDSYS have compiled-in defaults if they are not set.  The use of
these small setuid-root programs greatly eases upgrades to layers.

	The pseudo-tty version also has a enhancements over the XT-driver
based versions of layers.  The enhancements to layers include the handling
of the configuration file and disposition of dead layers (where the host
resident shell has terminated and the layer is now inactive):  Under
pseudo-tty layers, the configuration file can now contain commands which
have semantics similar to hostagent calls (see the libwindows man page).
A new optional mode of operation was added to layers to automatically
delete a window when the controlling shell exits.  When you exit the shell
attached to a window, the window on the terminal is deleted.  On XT-driver
based layers the window would remain, but would be left inactive.  Another
option will force layers to automatically exit when the last window is
deleted.  All these enhancements can be easily turned off to mirror
XT-driver layers semantics.  For more information, see the layers manual
page.

	Especially on systems other than SunOS, most of the J* ioctls do
not work on pseudo-tty layers.  This requires some small changes to
programs which relied upon them under XT-driver layers.  See the beginning
of the libwindows man page for details of changes required to these programs.

DIFFERENCES FROM THE KEITH MULLER VERSION OF PSEUDO-TTY LAYERS
	For details of changes, see the HISTORY file.  These are the
highlights:
    1. Compiles on many System-V variants as well as BSD variants.
    2. Openchan() libwindows call returns a pseudo-tty instead of sockets.
	This allows commands running in another window to work more like
	commands in normal windows.
    3. Supports up to 252-byte packets on 730 terminals in regular XT.
    4. Supports network XT protocol on 730 terminals.
    5. Flow control added on agent calls to make sure all data from the
	pseudo-tty is drained before processing the agent call.  This
	solves problems when starting up multiple downloads either in
	from multiple windows or sequentially in the same window.
    6. Layers no longer needs to be setuid-root.
    7. The 'who' file only shows one logged in window, just like XT-driver
	layers.  A 'relogin' command is part of the package to change the
	window that shows up in the 'who' file.
    8. 'xts', 'xtt', and 'xtagent' commands added to the package.
    9. Libwindows works under both XT-driver layers and pseudo-tty layers
	in the same binary.
    10. A bug which caused lost packets, NAKS, and timeouts due to
	interrupted writes is fixed.
    11. Debugging messages can be selectively enabled at run time.

SOURCE INSTALLATION STEPS
	The makefile is set up to compile /usr/local/dmdlayers as the
default DMD/DMDSYS path.  The toplevel makefile can be changed to alter this
default path.  The source files can be placed anywhere regardless of the
setting for DMD in the makefile (they should NOT be in the same directory
as the binaries though as the make will step on itself).  A good place for
the source files is $DMD/src (or $DMD/src/ptlayers if you also have the
sdp630 software development package and install both under $DMD).

	The layers package is designed to be able to be installed either in
the same location as the 630 software development package or in a different
location.  If they are both in the same location, you should be able to
ignore the setting of the DMDSYS variable and just always use DMD.  If they
are in different locations, the DMDSYS variable points to layers installation
location and the DMD variable points to software development package
installation location.

	The binaries can be owned by any user EXCEPT for the 'movelogin'
and 'grantpt' (if it exists) programs in $DMD/lib/layersys.  If /etc/utmp
is world-writable or if pututxent() can be called by non-root users,
movelogin does not need to be setuid-root; otherwise, layers will not be
able to update the who file.  'Grantpt' needs to be setuid-root to enable
to change owner and access rights to the ptys.  These programs can also be
in $DMDSYS/lib/layersys (a common default for $DMDSYS is "/usr").

	The directory $DMD/xt or $DMDSYS/xt must exist and be mode 0777 for
layers to start up.  That directory is where the unix domain sockets (or named
pipes depending on a compile-time option) used in the hostagent system are
bound.  The layers driver and each program that uses libwindows must be
able to bind UNIX domain sockets (or create named pipes) in that directory.

A step by step procedure:
	1) cd to where you have placed the source.
	2) Check out the file include/lyrscfg.h.  If you are on one of the
	   supported machine types you shouldn't have to make any changes.
	   Otherwise, you may have to edit it to select the proper defines
	   for your system.
	3) The software follows standard makefile practice. To install
	   from scratch do
		make install
	   at the top level. 
	   There are three location-specifying parameters you may want to
	   add on the make command line:
		DMDSYS - the location for make to install into (defaults to
			$(DMD))
		DEFDMD - the default value for $DMD if it is not set at
			run-time (defaults to $(DMDSYS))
		DEFDMDSYS - the default value for $DMDSYS if it is not set
			at run-time (defaults to $(DEFDMD))
	   You must be able to write to your chosen DMDSYS.  If you run
	   the make as root the required support programs will be installed
	   setuid-root.
	   You can clean it all out with
		make clean
	4) Add the termcap in misc/MTG.termcap to /etc/termcap if you wish.
	   The 5620 termcap not work properly with the 630 MTG.
	5) Add the $DMDSYS/bin to your path in your shell startup file.
	6) See the ATT manuals for running layers (and the supplied manual
	   pages) and for getting started.
	7) Make sure that the directory $DMDSYS/xt is mode 777.
	8) Once the software is compiled, there are no references to the source
	   subtree so the source can be moved or removed.

    ADDITIONAL STEPS FOR NFS/RFS HOSTS SO CLIENTS CAN SUPPORT PSUEDO-TTY LAYERS

	9) If you are running on a nfs/rfs served machine and layers is on
	   the server you may have problems writing to the directory
	   $DMDSYS/xt FROM THE CLIENT.  If you DO NOT need to run this
	   software on the client, SKIP THIS STEP.  Otherwise on the SERVER
	   do:

		a) rmdir $DMDSYS/xt
		b) choose a directory that will be local to each machine,
		    for example /var/spool/xt, and make a symbolic link to
		    it.  For example,
		   	ln -s /var/spool/xt $DMDSYS/xt

	   On EACH CLIENT that will need to run the pseudo-tty layers 

		a) Make your chosen directory. For example,
			mkdir /var/spool/xt
		b) Make the directory mode 777. For example,
			chmod 0777 /var/spool/xt

PORTING 
	The first thing to do is to edit include/lyrscfg.h.  Read the
comments carefully and try to select the options that are right for your
system.  If there aren't enough options for your system, add more and
enable them for your system in lyrscfg.h.  The options may be applicable to
other machine types too so that's preferable to putting in machine-specific
defines in the rest of the source.

	If you need to add any compile options, set the EXTRACFLAGS
variable in the top level makefile.  This flag is intended for use to turn
on compatibility modes on your compiler if there is a way to do that,
and if you need them to get layers to compile.

	My goal is to have the same source code compile on all supported
machine types with a minimum amount of editing.  Sometimes it may be
necessary to set some EXTRACFLAGS but I'd like to keep that to a minimum.
If at all possible, the defines in include/lyrscfg.h should be selected
based on machine and/or operating system-specific defines (for example "i386"
for '386 based systems) so they can compile without editing.

	You can compile layers with the DEBUG code defined (see layers/
Makefile) for testing a port, but layers with the DEBUG code compiled in
should not be used for production use as it is significantly slower and a
lot more expensive to use.

PROBLEMS
	If you have problems with any of this software or need help in porting
feel free to contact me:

	David Dykstra
	dwd@graceland.ih.att.com

							November, 1993


HISTORY
This file contains brief descriptions of changes made to the pseudo-tty
layers in chronological order.  This file was created 4/19/93 so dates
before that time are approximate.  Unless otherwise noted, all changes were
made by Dave Dykstra.

1991
====
10/1	Keith Muller's version of layers retrieved from anonymous ftp ucsd.edu 
	at pub/dmdlayers.tar.Z.

12/1	Modified openchan() libwindows call to return a pseudo-tty instead of
	a socket; this supports starting a program in a window from another
	process such as "agent !".  Also removed openrwchan() as it is no
	longer necessary.

1992
====
2/1	Added calls to signal() at the end of signal handling routines to
	reset signal handlers in case layers is compiled with /usr/5bin/cc
	on a sun.

4/23	Strip off eigth bit of TIOCUCNTL calls because the R&D Unix dmdld
	had turned the uppermost bit on for some reason.

5/1	Addition of one second delay after switching to layers mode to
	allow time for the terminal to initialize.

7/1	Allowed layers to come up even if not setuid-root.  When not setuid-
	root, can't write /etc/utmp unless it is world writable, and can't
	change ownership of pseudo-ttys.

9/1	Strip off 8th bit when inquiring the terminal for it's escape
	sequence identifier.  This allows getting into layers no matter
	what the parity setting.  Also allowed for missing square bracket
	in the returned escape sequence in case the wproc terminal controls
	option is set to 8 bits.  Will print out the id strings in a
	readable format if they are not recognized.

10/1	No longer requires empty files in $DMD/lib/layersys for each terminal
	id string to get into layers.

12/1	Added NO_TIOCUCNTL to support machine types on which the TIOCUCNTL
	ioctl on pseudo-ttys is not supported.

12/9	Added getagentanswer() libwindows call to get access to the return
	buffer from an agent call.  This is used by the agent(1) program to
	display the return buffer in ascii for agent calls like A_ROMVERSION.

12/14	Added '-V' option to print out version.
	Made lyrscfg.h to contain defines for different operating system
	configurations.
	Added support for mips-based OS's, specifically StarServer-FT,
	thanks to many changes by John Ampe.  This OS is a combination of
	System V and BSD.
	One notable addition is USE_NAMED_PIPE to use named pipes to
	communicate with support programs rather than unix-domain sockets.
	Another is support for system V utmp file format, although it is
	known to be buggy.
	Changed default shell to /bin/sh instead of /bin/csh because csh is
	not available on all systems.
	Added support for AT&T SVR4, USE_PTMX.

1993
====
2/1	Added -e option to force on hex encoding mode the same way that
	the environment variable DMDLOAD=hex does.

3/1	Increased number of pseudo-ttys supported from 96 to 256 on
	machines using the original bsd naming conventions
	(/dev/[pt]ty[pqrstuvwxyzPQRSTU][0-f]).

	Change to pass unrecognized agent calls on to the terminal rather
	than rejecting them immediately.

3/10	Added more portability defines for HP-PA thanks to Mark Larsen.

4/19	Added beginning of support for Solaris from Bob Pietkivitch.
	Added John Lash's '-m' option for larger packet size support on 730s.
	Got to compile without warnings on NCR 3000 SVR4 with both the AT&T
	and MetaWare compilers.
	Removed old sunos pty driver patches from misc directory.

4/21	Change to always return the result from an agent call in the agent
	answer buffer.
	Add flow control on agent calls to make sure all data from the
	pseudo-tty is drained before processing the agent call.
	Add support for A_JBOOT, A_JTERM, and A_JZOMBOOT even if on a system
	that has TIOCUCNTL.  The purpose is to be able to use the above
	flow control and to provide for transition to operating system
	versions that don't support TIOCUCNTL (such as sun -> solaris).
	Change libwindows to send the $DMDCHAN with the request if set
	before looking up ttyname() because it's much faster.

4/22	Add matchtty() call to match pty name on major/minor number if a
	match is not found by name.  This is for systems that have two /dev
	names for the same ptys: the one that layers sees may not be the
	same one that ttyname() sees, and having layers() call ttyname() is
	much slower.
	Change to print error message to opened window if can't allocate
	pty; previously the window was silently deleted.
	Made a more reasonable algorithm for determining timeout times
	based on the baud rate.  If the user does not specify '-b' baud
	rate, the maximum is assumed to be 4800 baud to allow for network
	delays and other delays; the timeout for 4800 baud is now 6 seconds
	for transmit and 4 seconds for receive, double the preceding minimum
	(and the mimimum was usually selected).
	Eliminated PXTIMEOUT and RXTIMEOUT defines and instead use the value
	for 1200 baud in the table at initialization time.
	Put in a one second delay when forking a process to create a new
	window to allow time for the process to open the pty before layers
	tries to read it.

4/27	Ignore leading blanks in a '-f' file.
	Fixed a problem with starting up multiple dmdld's from a '-f' file
	(only for libwindows programs that use A_JBOOT, not those using the
	JBOOT TIOCUCNTL ioctl): too many packets get sent on channel 0
	before being acked so some of them fail allocating a buffer to send
	the JBOOT message.  The fix is to delay agent messages if there are
	unacked messages on channel 0, using the agentblock flag.

4/28	Got rid of wtinit.  It was just a simple program that sent the
	startup string for layers, with hex encoding on or off.

5/4	Removed setenv.c and getenv.c.  Just use standard putenv/getenv instead.

5/5     Created 'grantpt' and 'movelogin' library programs.  These are small
	programs that are to be installed setuid root to grant ownership
	of pseudo ttys and to move entries around in the utmp file, respect-
	ively.  Now layers itself no longer needs to be setuid root.
	'Grantpt' is not needed on SVR4 because a similar program comes

layers/main.c
/*
 *	Copyright (c) 1987 Keith Muller
 *	All Rights Reserved
 *
 *	The author make no claims as to the fitness or correctness of
 *	this software for any use whatsoever, and it is provided ``as is''
 *	without express or implied warranty. Any use of this software is
 *	at the user's own risk.
 *
 *	Enhancements by David Dykstra at AT&T.
 *	This source code may be freely distributed.
 */
#include "common.h"
#include <sys/types.h>
#include <sys/time.h>
#include <sys/param.h>
#include <sys/file.h>
#include <fcntl.h>
#ifdef USE_UNISTD
#include <unistd.h>
#endif
#ifdef USE_TTYDEV
#include <sys/ttydev.h>
#endif
#include <stdio.h>
#include <signal.h>

#include "sys/jioctl.h"

#include "proto.h"
#include "packets.h"
#ifdef USE_TERMIOS_H
#include <sys/termios.h>
#endif
#ifdef USE_TERMIO_H
#include <sys/termio.h>
#endif
#if (defined(FAST) && !defined(USE_NICE))
#include <sys/resource.h>
#endif

char * Version = "pseudo-tty Keith Muller + Dave Dykstra 2/11/98";

extern struct layer layer[];
extern int fdtolayer[];

extern int Loadtype;
extern int regularxtarg;
extern int maxpktarg;
extern int highwaterarg;
extern int numpcbufsarg;
extern int nestedlayers;

extern int DebugLevel;
extern FILE *DebugFile;

static int Ppstatistics = 0;
static int Pptrace = 0;

extern char *getenv();

#ifndef DEFDMDSYS
#define DEFDMDSYS "/usr"
#endif

main(argc, argv)
int argc;
char **argv;
{
	register int i;
	extern void catchsig();
	extern char *ttyname();
	extern char asockname[];
	extern char fakespeed;
	extern int booted;
	char *tty;
#	ifdef FAST
	extern int oldpriority;
	extern int fast;

#ifdef USE_NICE
	if (nice(fast) != -1)
		oldpriority = -fast;
	else
		/* nice fails if fast is negative and not super-user */
		oldpriority = 0;
#else
	oldpriority = getpriority(PRIO_PROCESS, 0);
	(void) setpriority(PRIO_PROCESS, 0, oldpriority + fast);
#endif
#	endif

	setuid(getuid()); /* no more reason to be setuid-root */
	setgid(getgid());

	if (parse(argc, argv))
		exit(1);

	if (!nestedlayers && (getenv(AGENV) != NULL)) {
		fprintf(stderr,
			"Layers: attempting to invoke layers within a layer");
		fprintf(stderr, " ($%s is set).\n", AGENV);
		fprintf(stderr, " Use '-l' if that is what you really want.\n");
		exit(1);
	}

#ifdef USE_SYSV_SIGNALS
	sighold (SIGHUP);
	sighold (SIGTERM);
#ifdef SIGTSTP
	sighold (SIGTSTP);
#endif
#else
	(void) sigblock(sigmask(SIGHUP)|sigmask(SIGTERM)|sigmask(SIGTSTP));
#endif
	
	layer[0].flags = NONETFLOW;
	for (i = 0; i < MAXPCHAN; i++) {
		layer[i].fd = -1;
		layer[i].pgrp = -1;
	}
	fdtolayer[0] = 0;
	for (i = 1; i < MAXFD; i++)
		fdtolayer[i] = -1;

	if ((tty = ttyname(0)) == (char *)0) {
		fprintf(stderr, "Layers: stdin is not a tty.\n");
		exit(1);
	}

	initutmp(tty);

	savemode(0);

	if (initagent(tty)) {
		fprintf(stderr, "Layers: Problems starting agent system.\n");
		exit(1);
	}

	moveutmp(1);

	setcntrl(0, tty);
#if defined(USE_SYSV_SIGNALS) || defined(USE_SYSV_WAIT)
	(void) sigset(SIGCHLD, SIG_DFL);
#else
	(void) signal(SIGCHLD, SIG_IGN);
#endif
	if (boot(0)) {
		resetcntrl(0, tty);
		(void) unlink(asockname);
		fprintf(stderr, "Layers: Terminal boot failed, Exiting.\n");
		moveutmp(0);
		exit(1);
	}

	/* Wait until after boot to set agent environment because boot */
	/* could call 32ld and we don't want it to think that it is in layers */
	initagentenv();

	mysleep(3); /* give time for terminal to initialize */
		  /* 630 only needs 1 second but 5620 needs 3 */

	booted = 1;

	setcntrl(0, tty);
#ifndef USE_SYSV_SIGNALS
	(void) signal(SIGHUP, catchsig);
	(void) signal(SIGTERM, catchsig);
	(void) signal(SIGCHLD, catchsig);
	(void) signal(SIGTSTP, SIG_IGN);
	(void) signal(SIGPIPE, SIG_IGN);
#else
	(void) sigset(SIGHUP, catchsig);
	(void) sigset(SIGTERM, catchsig);
	(void) sigset(SIGCHLD, catchsig);
	(void) sigset(SIGPIPE, SIG_IGN);
#endif

	if (protocolinit(0)) {
		(void) sendioctl(0, JTERM_CHAR);
		resetcntrl(0, tty);
		(void) unlink(asockname);
		fprintf(stderr, "Layers: Cannot start protocol, exiting.\n");
		exit(1);
	}
#ifdef USE_SYSV_SIGNALS
	sigrelse (SIGHUP);
	sigrelse (SIGTERM);
#else
	(void) sigsetmask(0);
#endif
	multiplex();
	quit(0);
}

quit(reset)
int reset;
{
	register int n;
	extern void flush_write();

#ifdef USE_SYSV_SIGNALS
	sighold (SIGHUP);
	sighold (SIGTERM);
#ifdef SIGTSTP
	sighold (SIGTSTP);
#endif
#else
	(void)sigblock(sigmask(SIGHUP)|sigmask(SIGTERM)|sigmask(SIGTSTP));
#endif

	for (n = 0; n < MAXPCHAN; n++)
		deletehost(n);

	exitagent();
	if (!reset)
		(void) sendioctl(0, JTERM_CHAR);
	flush_write();
	alarm(0);
	resetcntrl(0, ttyname(0));
	moveutmp(0);
	if (DebugLevel > 0)
		fpsummary(DebugFile);
	if (Pptrace)
		fpdumphist(stderr);
	if (Ppstatistics)
		fpsummary(stderr);
	exit(0);
}

usage()
{
fprintf(stderr,
"Usage: layers [-V] [-xrstl] [-m max-pkt] [-h high-water] [-b baud] \\\n");
fprintf(stderr,
"		[-f startupfile] [-d debuglevel] [-o debugfile] [lsys]\n");
fprintf(stderr,
" -V : print out version number and exit\n");
fprintf(stderr,
" -x : use hex encoding regular XT (same as DMDLOAD=hex)\n");
fprintf(stderr,
" -r : use regular XT, not network XT (same as DMDLOAD=regular)\n");
fprintf(stderr,
" -s : send protocol summary to stderr on exit (similar to xts)\n");
fprintf(stderr,
" -t : send protocol trace to stderr on exit (similar to xtt)\n");
fprintf(stderr,
" -l : attempt running layers even in a layers window (most terminals don't)\n");
fprintf(stderr,
" -m max-pkt: max packet size for regular XT (same as DMDMAXPKT=max-pkt)\n");
fprintf(stderr,
" -h high-water: high water for network XT (same as DMDHIGHWATER=high-water)\n");
fprintf(stderr,
" -b baud: baud rate to use for determining regular XT timeouts\n");
fprintf(stderr,
" -f startupfile: layers startup file (default $HOME/.layersrc)\n");
fprintf(stderr,
" -d debuglevel : enable debugging. debuglevel is the OR of the following:\n");
fprintf(stderr,
"	1 : non-protocol errors\n");
fprintf(stderr,
"	2 : low-volume information messages\n");
fprintf(stderr,
"	4 : protocol errors\n");
fprintf(stderr,
"	8 : protocol trace\n");
fprintf(stderr,
"	16 : raw bytes to/from terminal\n");
fprintf(stderr,
"	32 : high-volume information messages\n");
fprintf(stderr,
" -o debugfile : file to use for debug messages (default \"layers.debug\")\n");
fprintf(stderr,
" lsys : layersys file to download before starting protocol.  5620 only.\n");
}

parse(argc, argv)
int argc;
char **argv;
{
	register int retval;
	register char *p;
	char *debugfname = "layers.debug";
	extern char fakespeed;
	extern char ffile[];
	extern char dmdpath[];
	extern char dmdsyspath[];
	extern char shell[];
	extern char lsys[];
	extern char *optarg;
	extern int optind;
	extern char *strncpy();
	extern int ptracing;

	Loadtype = UNKNOWN_LOAD;
	if ((p = getenv("DMDLOAD")) != (char *)0) {
		if (strncmp(p, "hex", 3) == 0)
			Loadtype = HEX_LOAD;
		else if (strncmp(p, "regular", 7) == 0)
			regularxtarg = 1;
	}

	if ((p = getenv("DMDMAXPKT")) != (char *)0) {
		if((p[0] < '0') || (p[0] > '9') ||
			((maxpktarg = atoi(p)) < MAXPKTDSIZE) ||
					(maxpktarg > MAXOUTDSIZE)) {
			fprintf(stderr,
			      "Invalid DMDMAXPKT %d; should be %d to %d\n",
				maxpktarg, MAXPKTDSIZE, MAXOUTDSIZE);
			exit(1);
		}
	}

	if ((p = getenv("DMDHIGHWATER")) != (char *)0) {
		if((p[0] < '0') || (p[0] > '9') ||
			((highwaterarg = atoi(p)) < MINNETHIWAT) ||
					(highwaterarg > MAXNETHIWAT)) {
			fprintf(stderr,
			      "Invalid DMDHIGHWATER %d; should be %d to %d\n",
				highwaterarg, MINNETHIWAT, MAXNETHIWAT);
			exit(1);
		}
	}
	while ((retval = getopt(argc, argv, "Vxrstlf:b:h:m:n:d:o:")) != EOF) {
		switch (retval) {
		case 'V':
			fprintf(stderr, "layers version: %s\n", Version);
			exit(0);
			break;
		case 'x':
			/* encoding mode forced on */
			Loadtype = HEX_LOAD;
			break;
		case 'r':
			/* regular XT (network xt forced off */
			regularxtarg = 1;
			break;
		case 's':
			/* print protocol statistics on exit */
			Ppstatistics++;
			break;
		case 't':
			/* print protocol trace on exit */
			Pptrace++;
			ptracing = 1;
			break;
		case 'f':
			(void) strncpy(ffile, optarg, (MAXPATHLEN-1));
			break;
		case 'b':
			switch (atoi(optarg)) {
			case 19200:
				fakespeed = EXTA;
				break;
			case 9600:
				fakespeed = B9600;
				break;
			case 4800:
				fakespeed = B4800;
				break;
			case 2400:
				fakespeed = B2400;
				break;
			case 1200:
				fakespeed = B1200;
				break;
			case 300:
				fakespeed = B300;
				break;
			default:
				fakespeed = 0;
				break;
			}
			break;
		case 'h':
			if((optarg[0] < '0') || (optarg[0] > '9') ||
				((highwaterarg = atoi(optarg)) < MINNETHIWAT) ||
					    (highwaterarg > MAXNETHIWAT)) {
				fprintf(stderr,
				  "Invalid high water %d; should be %d to %d\n",
					highwaterarg, MINNETHIWAT, MAXNETHIWAT);
				exit(1);
			}
			break;
		case 'm':
			if((optarg[0] < '0') || (optarg[0] > '9') ||
				((maxpktarg = atoi(optarg)) < MAXPKTDSIZE) ||
						(maxpktarg > MAXOUTDSIZE)) {
				fprintf(stderr,
				      "Invalid maxpkt %d; should be %d to %d\n",
					maxpktarg, MAXPKTDSIZE, MAXOUTDSIZE);
				exit(1);
			}
			break;
		case 'n':
			/* 
			 * This is not documented because it is dangerous;
			 * if 4 is chosen and the max pkt size > (252/2) and
			 *   the terminal is not modified to support the
			 *   extra buffers, it will cause many NAKs.
			 */
			if((optarg[0] < '0') || (optarg[0] > '9') ||
				((numpcbufsarg = atoi(optarg)) < NPCBUFS) ||
						(numpcbufsarg > MAXNPCBUFS)) {
				fprintf(stderr,
				      "Invalid numpcbufs %d; should be %d to %d\n",
					numpcbufsarg, NPCBUFS, MAXNPCBUFS);
				exit(1);
			}
			break;

		case 'l':
			nestedlayers++;
			break;

		case 'd':
			DebugLevel = atoi(optarg);
			break;
		
		case 'o':
			debugfname = optarg;
			break;
			
		default:
			usage();
	    		return(-1);
		}
	}

	DebugInit(debugfname);

	if ((DebugLevel & (DEBPTRACE|DEBPERROR)) != 0)
		ptracing = 1;

	if (optind < argc)
		(void) strncpy(lsys, argv[optind], (MAXPATHLEN-1));
	else
		lsys[0] = '\0';
	if ((p = getenv("SHELL")) == (char *)0 || *p == '\0')
		p = DSHELL;
	(void) strncpy(shell, p, (MAXPATHLEN-1));
	if ((p = getenv("DMD")) == (char *)0 || *p == '\0')
		p = DEFDMD;
	(void) strncpy(dmdpath, p, (MAXPATHLEN-1));
	if ((p = getenv("DMDSYS")) == (char *)0 || *p == '\0')
		p = DEFDMDSYS;
	(void) strncpy(dmdsyspath, p, (MAXPATHLEN-1));
	return(0);
}
